rslib/src/scheduler/states/load_balancer.rs

```diff
@@ -0,0 +1,196 @@
+    }
+}
+
+pub struct LoadBalancerContext<'a> {
+    load_balancer: &'a LoadBalancer,
+    note_id: Option<NoteId>,
+    fuzz_seed: Option<u64>,
+}
+
+impl<'a> LoadBalancerContext<'a> {
+    pub fn find_interval(&self, interval: f32, minimum: u32, maximum: u32) -> Option<u32> {
+        self.load_balancer
+            .find_interval(interval, minimum, maximum, self.fuzz_seed, self.note_id)
+    }
+
+    pub fn set_fuzz_seed(mut self, fuzz_seed: Option<u64>) -> Self {
+        self.fuzz_seed = fuzz_seed;
+        self
+    }
+}
+
+#[derive(Debug)]
+pub struct LoadBalancer {
+    days: [LoadBalancerDay; LOAD_BALANCE_DAYS],
+}
+
+impl LoadBalancer {
+    pub fn new(today: u32, deck_id: DeckId, storage: &SqliteStorage) -> Result<LoadBalancer> {
+        let cards_on_each_day =
+            storage.get_all_cards_due_in_range(today, today + LOAD_BALANCE_DAYS as u32)?;
+        let decks = storage.get_all_decks()?;
+
+        let current_deck_config_id = decks
+            .iter()
+            .find(|deck| deck.id == deck_id)
+            .and_then(|deck| deck.config_id());
+
+        let cards = match current_deck_config_id {
+            Some(current_deck_config_id) => {
+                let decks_in_preset = decks
+                    .iter()
+                    .filter(|deck| deck.config_id() == Some(current_deck_config_id))
+                    .map(|deck| deck.id)
+                    .collect::<Vec<_>>();
+
+                cards_on_each_day
+                    .into_iter()
+                    .map(|day| {
+                        day.into_iter()
+                            .filter(|(_, _, did)| decks_in_preset.contains(did))
+                            .collect()
+                    })
+                    .collect()
+            }
+            None => cards_on_each_day,
+        };
+
+        let mut days = std::array::from_fn(|_| LoadBalancerDay::default());
+        for (cards, cache_day) in cards.iter().zip(days.iter_mut()) {
+            for card in cards {
+                cache_day.add(card.0, card.1);
+            }
+        }
+
+        Ok(LoadBalancer { days })
+    }
+
+    pub fn review_context(&self, note_id: Option<NoteId>) -> LoadBalancerContext {
+        LoadBalancerContext {
+            load_balancer: self,
+            note_id,
+            fuzz_seed: None,
+        }
+    }
+
+    fn find_interval(
+        &self,
+        interval: f32,
+        minimum: u32,
+        maximum: u32,
+        fuzz_seed: Option<u64>,
+        note_id: Option<NoteId>,
+    ) -> Option<u32> {
+        // if we're sending a card far out into the future, the need to balance is low
+        if interval as u32 > MAX_LOAD_BALANCE_INTERVAL as u32
+            || minimum > MAX_LOAD_BALANCE_INTERVAL as u32
+        {
+            println!(
+                "load balancer: interval/minimum {}/{} over threshold {}, not balancing",
+                interval, minimum, MAX_LOAD_BALANCE_INTERVAL
+            );
+            return None;
```

Comment:
The other debugging statements have been removed; please remove this too, or use debug!() instead (set RUST_LOG=anki=debug to see it)

Reply:
totally missed that one